### DISCUSSION TITLE
Heap histogram doesn't work in Spark UI

### DISCUSSION THREAD

**Discussion by paulpaul1076** [stackabletech/spark-k8s-operator#340] [Q&A] [answered]
Hello, I saw that there's a new option in Spark UI, called heap histogram.

![image](https://github.com/stackabletech/spark-k8s-operator/assets/4533296/b5b132b6-dff1-42ee-9704-5af11b0bda97)

But when I click on it I get this:

![image](https://github.com/stackabletech/spark-k8s-operator/assets/4533296/572eaffb-0fac-4a16-bda4-6b71294d9035)

In executor logs I get this:

```
java.io.IOException: Cannot run program "/usr/lib/jvm/java-11-openjdk-11.0.21.0.9-2.el8.x86_64/bin/jmap": error=2, No such file or directory
    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1128) ~[?:?]
    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1071) ~[?:?]
    at org.apache.spark.util.Utils$.getHeapHistogram(Utils.scala:2147) ~[spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.storage.BlockManagerStorageEndpoint$$anonfun$receiveAndReply$1.applyOrElse(BlockManagerStorageEndpoint.scala:82) ~[spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:103) [spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213) [spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100) [spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75) [spark-core_2.12-3.5.0.jar:3.5.0]
    at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41) [spark-core_2.12-3.5.0.jar:3.5.0]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
    at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.io.IOException: error=2, No such file or directory
    at java.lang.ProcessImpl.forkAndExec(Native Method) ~[?:?]
    at java.lang.ProcessImpl.<init>(ProcessImpl.java:340) ~[?:?]
    at java.lang.ProcessImpl.start(ProcessImpl.java:271) ~[?:?]
    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1107) ~[?:?]
    ... 11 more
```

    
    
 How do we get this to work?
 
 I guess in order for this to work you have to rebundle a different version of the JDK that includes jmap?

**Comment by razvan** [comment, marked as answer]
Hey, yes you are correct.

It seems that UBI 8 only includes `jmap` in the `devel` version of the openjdk  package.

To build your own image with `jmap` included you could do something like this:

```
FROM docker.stackable.tech/stackable/spark-k8s:3.5.0-stackable23.11.0

USER root

RUN microdnf update && \
    microdnf install java-11-openjdk-devel && \
    microdnf clean all
```

To use your own Spark image, add the `custom` property to the application resource  like this:

```
apiVersion: spark.stackable.tech/v1alpha1
kind: SparkApplication
spec:
  sparkImage:
    custom: <your image name here>
    productVersion: "3.5.0"
    pullPolicy: IfNotPresent
...
```

Hope this helps.

**Reply by paulpaul1076** [reply]
We did this and it worked. Thanks! Do you plan to release next spark versions with openjdk-devel packaged in?

**Reply by sbernauer** [reply]
Hi @paulpaul1076 we have a weekly internal architectural discussion. I already added this exact question to the agenda, we will give you an update once we have a decision

**Reply by sbernauer** [reply]
The possible downside here being e.g. security (aiming for slim container images) and image sizes

**Reply by lfrancke** [reply]
I do think it's a valid use-case though and I'd be in favor of including it but let's see what we decide in our arch discussions.

**Reply by paulpaul1076** [reply]
After building my own docker image with microdnf looks like there's no more python and pyspark doesn't work, normal spark works:
Exception in thread "main" java.io.IOException: Cannot run program "/usr/bin/python": error=2, No such file or directory

**Reply by lfrancke** [reply]
FYI: We decided to include the `devel` versions in our images going forward.

https://github.com/stackabletech/docker-images/pull/537#discussion_r1466006812

**Reply by sbernauer** [reply]
Hi @paulpaul1076, we have added the needed tools in our current nightly version, they will be part of the next release.
Until than you can use the `0.0.0-dev` version, so that it pulls in the image  `docker.stackable.tech/stackable/spark-k8s:3.5.0-stackable0.0.0-dev`

A working example that just sleeps for a long time to access the WebUI is (the important part is the `stackableVersion`)
```
apiVersion: spark.stackable.tech/v1alpha1
kind: SparkApplication
metadata:
  name: pyspark-pi
  namespace: default
spec:
  version: "1.0"
  sparkImage:
    productVersion: 3.5.0
    stackableVersion: 0.0.0-dev # includes jmap
  mode: cluster
  mainApplicationFile: local:///stackable/spark/jobs/spark-script.py
  volumes:
    - name: script
      configMap:
        name: spark-script
  driver:
    config:
      volumeMounts:
        - name: script
          mountPath: /stackable/spark/jobs
  executor:
    replicas: 1
    config:
      volumeMounts:
        - name: script
          mountPath: /stackable/spark/jobs
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: spark-script
data:
  spark-script.py: |
    from pyspark.sql import SparkSession
    from pyspark.sql.types import StructType, StructField, StringType, LongType, ShortType, FloatType, DoubleType, BooleanType, TimestampType, MapType, ArrayType
    from pyspark.sql.functions import col, from_json, expr
    from datetime import datetime, timedelta
    import subprocess
    import time

    spark = SparkSession.builder.appName("spark-ingest-into-lakehouse").getOrCreate()
    time.sleep(99999)
```
![image](https://github.com/stackabletech/spark-k8s-operator/assets/29303194/f7938957-387b-49a7-a8b3-fb23b930d246)
I hope this helps!

**Reply by paulpaul1076** [reply]
@sbernauer @lfrancke  thank you very much! @sbernauer I see that this image has the "-dev" suffix, is it stable enough to be used for critical things? Or does it include some stuff that may be less stable than the non-dev image?

**Reply by lfrancke** [reply]
Well.... It is our nightly image so it can change at any time and we would not officially recommend using it in production yet. But we haven't introduced anything breaking yet and if it works for you I'd be pretty confident that you can use it. I recommend pinning a hash as the tag itself is floating.
Our next release is planned for March.

I'm not entirely sure for how long we keep nightly images around right now but I'm sure one of the others can answer that.

**Reply by paulpaul1076** [reply]
I ran this image and got some weird exception when expiring iceberg table snapshots:

```
org.apache.spark.SparkException: EOFException occurred while reading the port number from pyspark.daemon's stdout.
```

Haven't had the time to understand why it happens, will try to do it tomorrow and also try the stable image with python installed. I just need to install python with microdnf, right?

**Reply by razvan** [reply]
Hey, Python is already installed in the Stackable image and there is a[ link to it that is set explicitely](https://github.com/stackabletech/docker-images/blob/a51da74d74aab3f8510c131f5f45d5cd32110976/spark-k8s/Dockerfile#L37).
The problem seems to be that `microdnf update` also upated python it's self and removed that link.
It should be enough to restore it.

**Reply by paulpaul1076** [reply]
Thanks, now the dev image works, actually, seems like there was some other problem that resolved itself.
